### ISSUE TITLE
Question about agent on device with sporadic internet access

### ISSUE THREAD

**Issue by siredmar** [k3s-io/k3s#2596]
Hi!

Currently we are facing a problem we need to find a solution for. Please see the attached diagram below.
Note: This is purely theoretical and has not been implemented yet. Currently i am researching for solutions for our target architecture and need some expert input.

What architecture would you suggest i use? Maybe i should use a multicluster environment where every node acts as a own cluster? I am familiar with the project KubeEdge (https://github.com/kubeedge/kubeedge) that tries to make cloud computing for edge devices available acting as a bridge between the device and the kubernetes cluster. I evaluated kubeedge, but it seems far from production readyness.

So this is an approach for a target architecture:

![image](https://user-images.githubusercontent.com/2737320/100206029-1e530a80-2f06-11eb-9a80-e111dd14b256.png)

Explanation:
In this scenario there is a k3s master running somewhere in the cloud. There are several worker nodes (embedded Linux, k3s agents) that run several pods and services. There is a 3rd party communication hub for transport of user specific data. On cloud side there is another pod running, consuming this data from this 3rd party comm. hub. 

Problem:
As long as all componentes are connected everything works. The problem is that the edge nodes don't have a stable network connection. The 3rd party comm. hub will buffer all ongoing traffic and redirects them to the cloud part when connection is reastablished. However i guess the management of kubernetes resources won't work. 

I can guess some limitations here when connection to master is down:

- If a pod gets unhealthy (e.g. containers crashes) it won't be recreated
- If the device reboots for some reason the containers/pods won't be created
- DNS on same node won't work properly

Are there any more limitations i don't see right now or are false assumptions?

What i already tested is cutting down the node's connection to master. The pods seem to continue working (they all ran on the same node, communication only to each other).

**So basically my question is: Is there something on k3s side that acts as a minimalist 'master buffer' between the server and the agent?**
![image](https://user-images.githubusercontent.com/2737320/100209987-c4a10f00-2f0a-11eb-8145-a8a0ab19bd4d.png)

The current limitation with resource management in the temporary offline case right now is the only thing that blocks us from pursuing with k3s.

**Comment by brandond** [comment]
Kubernetes doesn't really have a good pattern for this; the kubelet is assumed to be in constant contact with the apiserver. The model most folks currently use is to put the entire cluster at the edge, and manage all the clusters with a tool like rancher or fleet.
